text/0038-protocol-api.md

```diff
@@ -0,0 +1,92 @@
+- Add as byproduct the possibility to add plugin with new protocol and own output (useless without RFC 39)
+
+# Motivation
+
+Create a better management of outputs, encoders and services with a protocol centered approach.
+
+Byproduct: Add also the possibility to support third-party plugin which add protocol support with his own output.
+
+*Services addition is a part of [RFC 39](0039-service-overhaul.md).*
+
+# Design
+*This RFC does not consider that because a protocol support a encoder it should be implemented in the main project.*
+
+## Protocols
+Add protocol as a new registerable object type in the API. This object will only provide at least his "official" acronym (ex: RTMP, WebRTC) and a list of all supported codecs. Except RTMP, it will only support H.264 and AAC for now.
+
+Protocol types shall be seen as their "specification" put as code, those are present to tell at least which codecs can be used with the protocol maybe even their protocol link prefix (`rtmp://`, `srt://`).
+It does not mean that services are compatible with all the codecs present in the specs. And neither mean that OBS support those codecs.
+
+Protocol like SRT are codec agnostic, so adding bool attributes telling that the protocol support any video and audio codecs could be a thing.
+
+Making it separate from outputs and services will make it usable for multi-protocol service from RFC 39 to check if a protocol is registered.
+
+The actually supported protocol with their specification documents :
+- RTMP (https://wwwimages2.adobe.com/content/dam/acom/en/devnet/rtmp/pdf/rtmp_specification_1.0.pdf)
+- RTMPS (same as RTMP)
+- HLS (https://datatracker.ietf.org/doc/html/rfc8216)
+- SRT (https://github.com/Haivision/srt/files/2489142/SRT_Protocol_TechnicalOverview_DRAFT_2018-10-17.pdf ; TLDR: codec agnostic)
+- FTL (no official one found except this https://github.com/microsoft/ftl-sdk/blob/master/README.md)
+
+## Services
+Adding to the API, the required protocol attribute just a char with the protocol name which tell OBS which protocol the service uses.
+
+With RFC 39, OBS could prevent showing services where the protocol is not registered or has no available output compatible with it.
+And also make services able to not show unsupported protocol.
+
+We shall consider that if a service handle a protocol, it doens't mean it can handle all the compatible codecs.
+So if a service can't handle all the compatible codec, adding a list of supported codec to a service related to his protocol shall be considered.
+
+`rtmp-services` will be modified to obtain this behavior beforehand with only H.264, AAC and Opus as compatible codecs.
+
+## Outputs
+Adding to the API, the required¹ protocol attribute just a char with the protocol name which tell OBS with which protocol it can be used.
+Also remove the compatible codecs attributes because it is a part of protocol now.
+And output should be able to support any codec that the protocol.
+
+A output only support one protocol.
```

Comment:
Does this limitation make sense?

Do you have to have a nvenc hls and an nvenc rtmp output now?

Reply:
> Does this limitation make sense?

An encoder provide an encoder for only one codec, so I guess a stream output provide an output for one protocol would make sense.

> Do you have to have a nvenc hls and an nvenc rtmp output now?

I think there is confusion between encoder and output objects.

I will try to add more precision to avoid confusion.

Reply:
This RFC was moved to #45.